### PR TITLE
Add viegenre cipher

### PR DESCRIPTION
**Pull request by sahilarora3117**

### CODE REVIEW

**Comment by arshubham** [comment]
@sahilarora3117 Working great. Just a few observations:
1. Can you handle the case if key length is 0. Currently when encoding or decoding it cycles through random outputs.
2. If possible add an option to preserve the case. I found an example online https://cryptii.com/pipes/vigenere-cipher (This needs to be done for other applicable ciphers too)

**Comment by sahilarora3117** [comment]
7d832d4 fixes point 1. Will make another PR fixes all the required ciphers

File: src/views/VigenereCipherView.vala [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+            encipher_button.valign = Gtk.Align.CENTER;
+
+            var box = new Gtk.Box (Gtk.Orientation.HORIZONTAL, 0);
+            box.margin_top = 8;
+            box.valign = Gtk.Align.CENTER;
+            box.pack_start (new Cipher.Widgets.Label (_("Key: ")), false, false, 0);
+            box.pack_start (key_entry, false, true, 0);
+
+            attach (new Cipher.Widgets.Label (_("Plain Text")), 0, 1, 1, 1);
```
Comment (arshubham) [inline src/views/VigenereCipherView.vala]: https://github.com/arshubham/cipher/blob/2118d35af8569c55642ee17477163c14483dfd96/src/views/VigenereCipherView.vala#L74-L81
```suggestion
            var grid = new Gtk.Grid ();

            grid.attach (new Cipher.Widgets.Label (_("Plain Text")), 0, 1, 1, 1);
            grid.attach (new Cipher.Widgets.ScrolledWindow (plaintext_textview), 0, 2, 1, 1);
            grid.attach (box, 0, 3, 1, 1);
            grid.attach (encipher_button, 0, 3, 1, 1);
            grid.attach (new Cipher.Widgets.Label (_("Cipher Text")), 0, 4, 1, 1);
            grid.attach (new Cipher.Widgets.ScrolledWindow (ciphertext_textview), 0, 5, 1, 1);
            grid.attach (decipher_button, 0, 6, 1, 1);

            attach (grid, 0, 0, 1, 1);
            attach (keynull_toast, 0, 0, 1, 1);
```

Comment (arshubham) [reply src/views/VigenereCipherView.vala]: This will fix the sliding issue.

File: src/views/VigenereCipherView.vala [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+
+        public VigenereCipherView () {
+            var vigenere = new Cipher.Ciphers.Vigenere ();
+
+            encipher_button.clicked.connect (() => {
+                ciphertext_textview.buffer.text =
+                vigenere.encrypt (plaintext_textview.buffer.text, key_entry.buffer.text);
+                if (key_entry.buffer.text == "") {
+                    keynull_toast.send_notification ();
```
Comment (arshubham) [inline src/views/VigenereCipherView.vala]: return if key is empty

```suggestion
                    keynull_toast.send_notification ();
                    return;
```

File: src/views/VigenereCipherView.vala [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+                }
+            });
+
+            decipher_button.clicked.connect (() => {
+                plaintext_textview.buffer.text =
+                vigenere.decrypt (ciphertext_textview.buffer.text, key_entry.buffer.text);
+                if (key_entry.buffer.text == "") {
+                    keynull_toast.send_notification ();
```
Comment (arshubham) [inline src/views/VigenereCipherView.vala]: Same here

File: src/views/VigenereCipherView.vala [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            });
+        }
+
+        construct {
+            plaintext_textview = new Cipher.Widgets.TextView ();
+            ciphertext_textview = new Cipher.Widgets.TextView ();
+            key_entry = new Cipher.Widgets.Entry ();
+            key_entry.editable = true;
+            key_entry.valign = Gtk.Align.CENTER;
+            keynull_toast = new Granite.Widgets.Toast (_("Key Entry is empty    "));
```
Comment (arshubham) [inline src/views/VigenereCipherView.vala]:
```suggestion
            keynull_toast = new Granite.Widgets.Toast (_("Key must be at least 1 character long"));
```

**Pull request merged**